### PR TITLE
Surface reactions

### LINKED ISSUE #727: Recombination/Dissociation flux BCs
## Recombination and dissociation flux BCs in a multi-species context

Assume you have 2 species A and B, associated with concentrations $c_A$ and $c_B$ in the bulk (in m-3).

At a surface in contact with a gas, you can have these three 2nd order reactions

A + A <-> A2
A + B <-> AB
B + B <-> B2

Species A2, AB, and B2 are associated with partial pressures $P_{A2}$, $P_{AB}$ and $P_{B2}$.

The reactions are associated with forward rate constants $K_{rAA}$, $K_{rAB}$ and $K_{rBB}$ in $\mathrm{m^{4} \ s^{-1}}$, and backwards rate constants $K_{dA2}$, $K_{dAB}$ and $K_{dB2}$ expressed in $\mathrm{m^{-2} \ s^{-1} \ Pa^{-1}}$.

The flux of species A and B at the surface is therefore

$$
\varphi_{A} = -2 K_{rAA} \ c_A \ c_A + 2 K_{dA2} P_{A2} - K_{rAB} \ c_A \ c_B + K_{dAB} P_{AB}
$$

$$
\varphi_{B} =- 2 K_{rBB} \ c_B \ c_B + 2 K_{dB2} P_{B2} - K_{rAB} \ c_A \ c_B + K_{dAB} P_{AB}
$$

## Simple implementation 

This is a simple design similar to `Reaction`. The product pressure here (eg. $P_{AB}$) is not dynamically calculated from the particle flux.

```python
import festim as F

A = F.Species(name="A")
B = F.Species(name="B")

surf_reaction_1 = F.SurfaceReactionBC(
    reactant=[A, A], product_pressure=2.0, k_r=0.1, k_d=1, surface=...
)
surf_reaction_2 = F.SurfaceReactionBC(
    reactant=[A, B], product_pressure=0.0, k_r=0.1, k_d=1, surface=...
)
surf_reaction_3 = F.SurfaceReactionBC(
    reactant=[B, B], product_pressure=1.0, k_r=0.1, k_d=1, surface=...
)
```

## Future proof implementation 

But what if we want to dynamically calculate the pressure of the gas species. Then maybe we create a new `GasSpecies` object? this would allow us later on to implement enclosures (see this [verification case](https://festim-vv-report.readthedocs.io/en/latest/verification/depleting_source.html)).

However, a full implicit implementation of this would require `Real` finite elements which are [not currently supported in dolfinx](https://github.com/FEniCS/dolfinx/issues/225).

```python

import festim as F

A = F.Species(name="A")
B = F.Species(name="B")

A2 = F.GasSpecies(name="A2", initial_pressure=2.0)
AB = F.GasSpecies(name="AB", initial_pressure=0)
B2 = F.GasSpecies(name="B2", initial_pressure=1.0)

surf_reaction_1 = F.SurfaceReactionBC(reactant=[A, A], product=[A2], k_r=0.1, k_d=1, surface=...)
surf_reaction_2 = F.SurfaceReactionBC(reactant=[A, B], product=[AB], k_r=0.1, k_d=1, surface=...)
surf_reaction_3 = F.SurfaceReactionBC(reactant=[B, B], product=[B2], k_r=0.1, k_d=1, surface=...)

```

### PR DESCRIPTION
**Pull request by RemDelaporteMathurin**
## Proposed changes

This PR fixes #727 
Needs to be merged after #752 

Also some refactoring. Got rid of `HTransportProblem.create_flux_value_fenics` and merged it with `define_boundary_conditions` (which breaks some of the tests). I happy to not do this here and keep it for another PR.

Todo:

- [x] Write tests
- [x] Write example and usage

## Types of changes

What types of changes does your code introduce to FESTIM?

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code refactoring
- [ ] Documentation Update (if none of the other choices apply)
- [ ] New tests

## Checklist

- [x] Black formatted
- [x] Unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

## Further comments

@jhdark I chose not to make this a special case of the very generic class ParticleFluxBC (_ie_ using a lambda function for value) to keep a simple implementation.

### CODE REVIEW

File: festim/hydrogen_transport_problem.py [hunk 1]
```diff
@@ -512,14 +511,30 @@ def define_meshtags_and_measures(self):
...
+        """Create forms for DirichletBC and value_fenics for ParticleFluxBC"""
+        # @jhdark this all_bcs could be a property
+        # I just don't want to modify self.boundary_conditions
+
+        # create all_bcs which includes all flux bcs from SurfaceReactionBC
+        all_bcs = self.boundary_conditions.copy()
         for bc in self.boundary_conditions:
+            if isinstance(bc, F.SurfaceReactionBC):
+                all_bcs += bc.flux_bcs
+                all_bcs.remove(bc)
```
Comment (RemDelaporteMathurin) [inline festim/hydrogen_transport_problem.py:523]: @jhdark here `all_bcs` could be a property I just don't want to modify `self.boundary_conditions`

**Comment by RemDelaporteMathurin** [comment]
@KulaginVladimir would you like to give this a look?

**Review by jhdark — APPROVED** [review]
Think this can be merged after pulling from the fenicsx branch. Although will likely have to revisit after chemical potential integration anyhow

**Comment by RemDelaporteMathurin** [comment]
@jhdark I finally rebased with `fenicsx`. It will need adapting for the mixed domain approach for sure but right now works with `HTransportProblem` with one material. I'll merge this after the CI has finished

File: test/system_tests/test_surface_reactions.py [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+            F.Species(),  # just a dummy species here
+            reaction.subdomain,
+        )
+        self.reaction = reaction.flux_bcs[0]
+
+    def compute(self, ds):
+        self.value = fem.assemble_scalar(
+            fem.form(self.reaction.value_fenics * ds(self.surface.id))
+        )
+        self.data.append(self.value)
```
Comment (RemDelaporteMathurin) [inline test/system_tests/test_surface_reactions.py:21]: We should add this as a FESTIM feature

**Pull request merged**